### PR TITLE
block refactor, fixed key aes, cointoss

### PR DESCRIPTION
**Pull request by sinui0**
This PR refactors some of our core types, and adds a simple cointoss protocol.

Mostly its a few odds and ends from the in-progress OT work.

# Changes
1. Replace the internal type of `Block` to be a `[u8; 16]`. `u128` was a mistake, blocks shouldn't be interpreted as integers but as a bag of bits. Eventually we will want to replace this with a type providing vectorized operations where possible.
2. Introduces cointoss protocol, needed for KOS
3. Pulls out some of the methods on `Block` into `FixedKeyAes`.

### CODE REVIEW

File: mpz-core/src/lib.rs [hunk 1]
```diff
@@ -2,11 +2,19 @@
...
 pub mod utils;
 pub mod value;
 
 pub use block::{Block, BlockSerialize};
+
+/// A protocol with a message type.
+pub trait ProtocolMessage {
+    /// The type of message used in the protocol.
+    type Msg: Send + Sync + std::fmt::Debug + 'static;
+}
```
Comment (th4s) [inline mpz-core/src/lib.rs:20]: Do you intend to add methods to this trait at some later point?

File: mpz-core/src/cointoss.rs
Comment (th4s) [inline mpz-core/src/cointoss.rs:1]: Can we split that into `sender` and `receiver` modules?

File: mpz-core/src/cointoss.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+                actual: receiver_seeds.len(),
+            });
+        }
+
+        if commitments.len() != receiver_seeds.len() {
+            return Err(CointossError::CountMismatch {
+                expected: commitments.len(),
+                actual: receiver_seeds.len(),
+            });
+        }
```
Comment (th4s) [inline mpz-core/src/cointoss.rs]: Isn't this already checked in `reveal`?

**Review by th4s — APPROVED** [review]
Nice :rocket:

**Comment by themighty1** [comment]
I get compilation errors when running `cargo test`

File: mpz-core/src/lib.rs (hunk 1 above)
Comment (sinui0) [reply mpz-core/src/lib.rs:20]: No, this is simply to associate a message type with something implementing a protocol

**Comment by sinui0** [comment]
> I get compilation errors when running `cargo test`

This is expected, as this PR is scoped only to `mpz-core`. Following PRs will refactor dependents to accomodate these changes. You can test `mpz-core` in isolaton

**Comment by sinui0** [comment]
I've slightly modified the cointoss to commit to all seeds at once, this is more efficient.

File: mpz-core/src/aes.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+//! Fixed-key AES cipher
+
+use aes::{cipher::generic_array::GenericArray, Aes128};
+use cipher::{generic_array::functional::FunctionalSequence, BlockEncrypt, KeyInit};
+use once_cell::sync::Lazy;
+
+use crate::Block;
+
+/// AES fixed key
```
Comment (themighty1) [inline mpz-core/src/aes.rs]: /// A fixed AES key (arbitrarily chosen)

File: mpz-core/src/aes.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+/// Fixed-key AES cipher
+pub struct FixedKeyAes {
+    aes: Aes128,
+}
+
+impl FixedKeyAes {
+    /// Tweakable circular correlation-robust hash function instantiated
+    /// using fixed-key AES.
+    ///
+    /// See <https://eprint.iacr.org/2019/074>
```
Comment (themighty1) [inline mpz-core/src/aes.rs]: See <https://eprint.iacr.org/2019/074> (section 7.4)

File: mpz-core/src/aes.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    ///
+    /// `π(π(x) ⊕ i) ⊕ π(x)`
```
Comment (themighty1) [inline mpz-core/src/aes.rs]: /// `π(π(x) ⊕ i) ⊕ π(x)` , where π can be instantiated using AES with a fixed key.

File: mpz-core/src/block.rs [hunk 6]
```diff
@@ -1,35 +1,35 @@
...
 use core::ops::{BitAnd, BitXor};
-use itybity::{BitLength, GetBit, Lsb0, Msb0};
-use rand::{CryptoRng, Rng};
+use itybity::{BitIterable, BitLength, GetBit, Lsb0, Msb0};
+use rand::{distributions::Standard, prelude::Distribution, CryptoRng, Rng};
 use serde::{Deserialize, Serialize};
-use std::convert::{From, TryInto};
+use std::convert::From;
 
 /// A block of 128 bits
```
Comment (themighty1) [inline mpz-core/src/block.rs:9]: /// A block of 128 bits in big-endian byte order

File: mpz-core/src/cointoss/mod.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+//! A simple 2-party coin-toss protocol.
+//!
+//! # Example
```
Comment (themighty1) [inline mpz-core/src/cointoss/mod.rs:3]: Does this example serve as a test? How can i run it? Does the CI need to be updated to also run examples?

**Review by themighty1 — APPROVED** [review]
gw, ack
i added tests which cover some corner cases

File: mpz-core/src/block.rs (hunk 6 above)
Comment (sinui0) [reply mpz-core/src/block.rs:9]: This comment would leak internal details of how the bits are stored. The user should not be privy to such things, ie assume it in their code.

File: mpz-core/src/cointoss/mod.rs (hunk 7 above)
Comment (sinui0) [reply mpz-core/src/cointoss/mod.rs:3]: it is a doctest, it runs when you run `cargo test`. The CI will run it as it is now

File: mpz-core/src/block.rs (hunk 6 above)
Comment (themighty1) [reply mpz-core/src/block.rs:9]: But a user who instantiates `let b = Block::new(bytes)` and later calls b.lsb() must know in which order to place their `bytes`, otherwise they will get  the wrong lsb, no?

Comment (themighty1) [reply mpz-core/src/block.rs:9]: @sinui0 ---^

File: mpz-core/src/cointoss/mod.rs (hunk 7 above)
Comment (themighty1) [reply mpz-core/src/cointoss/mod.rs:3]: ok, I see. However, I noticed that vscode's code navigation doesnt work for doctests (lmk if it's just me). It creates unnecesary friction when examining the code. Can we duplicate the doctests also as normal tests? 
@sinui0

File: mpz-core/src/block.rs (hunk 6 above)
Comment (sinui0) [reply mpz-core/src/block.rs:9]: This is a good point. However, we should still avoid `endian` semantics, as that implies that the block is interpreted as an integer. I would opt to rename the `lsb` function to `color_bit` in this case. The user should not be privy to the location of the color bit in the underlying type.

File: mpz-core/src/cointoss/mod.rs (hunk 7 above)
Comment (sinui0) [reply mpz-core/src/cointoss/mod.rs:3]: I generally don't think code navigation should come into play for something like this. Ideally in the future we will have many many doctests, would we want to duplicate our entire test suite?

File: mpz-core/src/block.rs (hunk 6 above)
Comment (themighty1) [reply mpz-core/src/block.rs:9]: I agree with that direction. But using `color_bit` would leak the GC terminology. It can be called generically like e.g. `last_bit`or `rightmost_bit`

File: mpz-core/src/cointoss/mod.rs (hunk 7 above)
Comment (themighty1) [reply mpz-core/src/cointoss/mod.rs:3]: @sinui0 The first thing that I do when looking at how a protocol works is find an end-to-end test and from there to dive deep into it with code navigation. Not being able to do so would cause significant friction.

Until rust-analyzer supports navigation in doctests, we will have to duplicate those tests :(
